206. Reverse Linked List/SecondSolution.java

```diff
@@ -0,0 +1,37 @@
+public class SecondSolution {
+    // Time: O(n), Space: O(n), n は LinkedList の長さ
+    public ListNode reverseList(ListNode head) {
+        return reverseListHelper(null, head);
+    }
+
+    private ListNode reverseListHelper(ListNode prev, ListNode current) {
+        if (current == null) {
+            return prev;
+        }
+        ListNode next = current.next;
+        current.next = prev;
+        prev = current;
+        current = next;
+        return reverseListHelper(prev, current);
+    }
+
+    // 別の再帰の書き方
+    // Time: O(n), Space: O(n), n は LinkedList の長さ
+    public ListNode reverseList(ListNode head) {
+        return reverseListHelper(head);
+    }
+
+    private ListNode reverseListHelper(ListNode current) {
+        if (current == null) {
+            return null;
+        } else if (current.next == null) {
+            return current;
+        }
```

Comment:
@ryoooooory 
原則 if と else if の関心の対象は揃える方がいいと思います。

最初のif 句で注目しているのが`current`なのに、`else if`で注目してるのが`current.next`になっています。

結果的に問題はないのですが、if と else if で注目している対象が変わると認知コストが上がると思います。


あと、`else if`があるのに`elseが`無いのも不自然かなと

 

```suggestion
        if (current == null) {
            return null;
        } 
        if (current.next == null) {
            return current;
        }
```

この辺の違和感、もっと説得力のある説明ができるきがするのですが、うまく言葉が出てきません (- -; )

Reply:
```java
if (A) {
  X
} else if (B) {
  Y
}
Z
```
と else if があると、X の続きが Z にあることを期待しますね。

Reply:
ありがとうございます！たしかにelse ないことやこのケースは特にifを分離した方がいいかとおもいましたのでそちらで修正しました！

Reply:
IMO: 
遅ればせながら拝見しました．
私もelse ifブロックを作るよりは，ifブロックを二つ作るのが良いと思います．
（以前指摘をいただいたことがあります）

https://discord.com/channels/1084280443945353267/1218740927120674977/1243941990983274586